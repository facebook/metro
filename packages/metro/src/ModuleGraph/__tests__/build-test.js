/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+javascript_foundation
 * @format
 */

'use strict';

const build = require('../build');
const defaults = require('metro-config/src/defaults/defaults');

const FILE_TYPE = 'module';

describe('build setup', () => {
  const buildSetup = (entryPointPaths, options) =>
    build({
      entryPointPaths,
      getPolyfills: () => [],
      graphFn,
      optimize: false,
      platform: 'ios',
      postProcessModules: mds => {
        // TODO: the .sort() function returns -1 1 or 0 for sorting, not boolean
        return [...mds].sort((l, r) => l.file.path > r.file.path);
      },
      translateDefaultsPath: x => x,
      ...options,
    });
  const polyfillOptions = {getPolyfills: () => ['polyfill-a', 'polyfill-b']};
  const noOptions = {};
  const noEntryPoints = [];

  it('adds a prelude containing start time and `__DEV__` to the build', async () => {
    const result = await buildSetup(noEntryPoints, noOptions);

    const [prelude] = result.modules;
    expect(prelude).toMatchSnapshot();
  });

  it('sets `__DEV__` to false in the prelude if optimization is enabled', async () => {
    const result = await buildSetup(noEntryPoints, {optimize: true});
    const [prelude] = result.modules;
    expect(prelude).toMatchSnapshot();
  });

  it('places the module system implementation directly after the prelude', async () => {
    const result = await buildSetup(noEntryPoints, noOptions);
    const [, moduleSystem] = result.modules;
    expect(moduleSystem).toEqual({
      dependencies: [],
      file: {
        code: '',
        path: defaults.moduleSystem,
        type: FILE_TYPE,
      },
    });
  });

  it('places polyfills after the module system', async () => {
    const result = await buildSetup(noEntryPoints, polyfillOptions);
    const list = polyfillOptions.getPolyfills();
    const polyfills = result.modules.slice(2, list.length + 2);
    expect(polyfills).toEqual(list.map(moduleFromPath));
  });

  it('places all entry points and dependencies at the end, post-processed', async () => {
    const entryPoints = ['b', 'c', 'd'];
    const result = await buildSetup(entryPoints, noOptions);
    expect(result.modules.slice(-4)).toEqual(
      ['a', 'b', 'c', 'd'].map(moduleFromPath),
    );
  });
});

function moduleFromPath(path) {
  return {
    dependencies: path === 'b' ? ['a'] : [],
    file: {
      code: '',
      path,
      type: FILE_TYPE,
    },
  };
}

async function graphFn(entryPoints, platform, options, callback) {
  const modules = Array.from(entryPoints, moduleFromPath);
  const depModules = Array.prototype.concat.apply(
    [],
    modules.map(x => x.dependencies.map(moduleFromPath)),
  );
  return {
    entryModules: modules,
    modules: modules.concat(depModules),
  };
}
