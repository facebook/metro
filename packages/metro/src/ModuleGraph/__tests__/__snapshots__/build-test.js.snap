// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build setup adds a prelude containing start time and \`__DEV__\` to the build 1`] = `
Object {
  "dependencies": Array [],
  "file": Object {
    "code": "var __NUM_MODULES__=0,__FRAMEWORK__=undefined,__DEV__=true,__BUNDLE_START_TIME__=this.nativePerformanceNow?nativePerformanceNow():Date.now(),process=this.process||{};process.env=process.env||{};process.env.NODE_ENV=\\"development\\";",
    "libraryIdx": null,
    "map": null,
    "path": "/<generated>/prelude.js",
    "type": "script",
  },
}
`;

exports[`build setup sets \`__DEV__\` to false in the prelude if optimization is enabled 1`] = `
Object {
  "dependencies": Array [],
  "file": Object {
    "code": "var __NUM_MODULES__=0,__FRAMEWORK__=undefined,__DEV__=false,__BUNDLE_START_TIME__=this.nativePerformanceNow?nativePerformanceNow():Date.now(),process=this.process||{};process.env=process.env||{};process.env.NODE_ENV=\\"production\\";",
    "libraryIdx": null,
    "map": null,
    "path": "/<generated>/prelude.js",
    "type": "script",
  },
}
`;
